In [36]:
!ls "input/1850/GNA_bk_1850.txt"

input/1850/GNA_bk_1850.txt


In [37]:
#load data
import glob # import glob to retrieve files/pathnames matching a specified pattern
file_path = "input/1850/GNA_bk_1850.txt"
with open(file_path, 'r', encoding="cp1252") as f: # The encoding was different from the previous 2
  data = f.readlines()  

In [38]:
import pandas as pd
pd.DataFrame(data).shape

(17089, 1)

There are 17,089 rows

## Moving the [... down to the next row

In [39]:
data

['HE ARNES’\n',
 'BROOKLYN CITY\n',
 'DIRECTORY.\n',
 '1850-1851.\n',
 'NOTE.—Names having a * are the names of colored people\n',
 'Abbreviationsh. stands for house, n, for near, c. for corner, op. for\n',
 'opposite,b. for between, The precise location of residences thus de-\n',
 'scribed, may be ascertained by reference to the SHEET DIREOTORY.\n',
 'A\n',
 'Abberly Richard, shoemaker 168 Hudson av\n',
 'Abberly Samuel, mason Dean n Vanderbilt av\n',
 'Abbey A C, carpenter 82 Poplar\n',
 'ABBES HORATIO G, Columbia Institute for the education of\n',
 'boys 75 Columbia\n',
 'Abbey W S, merchant NYh Union n Court\n',
 'Abbott Abraham, marble sawyer Pacific n Boerum\n',
 'Abbott Daniel, 25 Front\t[Court\n',
 'Abbott Francis H, merchant 141 Front NYh Joralemon n\n',
 'Abbott H B, milkman Kent av north\n',
 'Abbott James, laborer Bond c State\t[Hoyt\n',
 'Abbott John D, merchant 295.5 Pearl NYh 362 Atlantic n\n',
 'Abbott Moses, furrier 74 Concord\n',
 'Abbott Samuel D, butcher Myrtle c Ja

In [40]:
import re
for i in range(0, 120):
    if len(re.findall(r"\[.*$", data[i])) == 1:
        
        
        
        
        
        data[i + 1] = data[i + 1].replace('\n', re.findall(r"\[.*$", data[i])[0])
        data[i] = data[i].replace(re.findall(r"\[.*$", data[i])[0], "") # Delete [... from the old row
    else:
        data[i] = data[i]

In [41]:
data

['HE ARNES’\n',
 'BROOKLYN CITY\n',
 'DIRECTORY.\n',
 '1850-1851.\n',
 'NOTE.—Names having a * are the names of colored people\n',
 'Abbreviationsh. stands for house, n, for near, c. for corner, op. for\n',
 'opposite,b. for between, The precise location of residences thus de-\n',
 'scribed, may be ascertained by reference to the SHEET DIREOTORY.\n',
 'A\n',
 'Abberly Richard, shoemaker 168 Hudson av\n',
 'Abberly Samuel, mason Dean n Vanderbilt av\n',
 'Abbey A C, carpenter 82 Poplar\n',
 'ABBES HORATIO G, Columbia Institute for the education of\n',
 'boys 75 Columbia\n',
 'Abbey W S, merchant NYh Union n Court\n',
 'Abbott Abraham, marble sawyer Pacific n Boerum\n',
 'Abbott Daniel, 25 Front\t\n',
 'Abbott Francis H, merchant 141 Front NYh Joralemon n',
 'Abbott H B, milkman Kent av north',
 'Abbott James, laborer Bond c State\t',
 'Abbott John D, merchant 295.5 Pearl NYh 362 Atlantic n',
 'Abbott Moses, furrier 74 Concord',
 'Abbott Samuel D, butcher Myrtle c Jay h 115 Johnson',
 'A

In [17]:
data

['HE ARNES’\n',
 'BROOKLYN CITY\n',
 'DIRECTORY.\n',
 '1850-1851.\n',
 'NOTE.—Names having a * are the names of colored people\n',
 'Abbreviationsh. stands for house, n, for near, c. for corner, op. for\n',
 'opposite,b. for between, The precise location of residences thus de-\n',
 'scribed, may be ascertained by reference to the SHEET DIREOTORY.\n',
 'A\n',
 'Abberly Richard, shoemaker 168 Hudson av\n',
 'Abberly Samuel, mason Dean n Vanderbilt av\n',
 'Abbey A C, carpenter 82 Poplar\n',
 'ABBES HORATIO G, Columbia Institute for the education of\n',
 'boys 75 Columbia\n',
 'Abbey W S, merchant NYh Union n Court\n',
 'Abbott Abraham, marble sawyer Pacific n Boerum\n',
 'Abbott Daniel, 25 Front\t\n',
 'Abbott Francis H, merchant 141 Front NYh Joralemon n \n',
 'Abbott H B, milkman Kent av north \n',
 'Abbott James, laborer Bond c State\t\n',
 'Abbott John D, merchant 295.5 Pearl NYh 362 Atlantic n \n',
 'Abbott Moses, furrier 74 Concord \n',
 'Abbott Samuel D, butcher Myrtle c Jay h 115

In [14]:
import re

for i in range(0, 100):
    bring_down = re.findall(r"\[.*$", data[i])[0]
    
    try:
        if len(bring_down) >= 1:
            x = str(data[i+1])
            data[i+1] = x.replace('\n', bring_down) + '\n'
            data[i] = data[i].replace(bring_down, "") # Delete [... from the old row
        
    except IndexError:
        pass
    
    return data

IndexError: list index out of range

In [7]:
data

['HE ARNES’\n',
 'BROOKLYN CITY\n',
 'DIRECTORY.\n',
 '1850-1851.\n',
 'NOTE.—Names having a * are the names of colored people\n',
 'Abbreviationsh. stands for house, n, for near, c. for corner, op. for\n',
 'opposite,b. for between, The precise location of residences thus de-\n',
 'scribed, may be ascertained by reference to the SHEET DIREOTORY.\n',
 'A\n',
 'Abberly Richard, shoemaker 168 Hudson av\n',
 'Abberly Samuel, mason Dean n Vanderbilt av\n',
 'Abbey A C, carpenter 82 Poplar\n',
 'ABBES HORATIO G, Columbia Institute for the education of\n',
 'boys 75 Columbia\n',
 'Abbey W S, merchant NYh Union n Court\n',
 'Abbott Abraham, marble sawyer Pacific n Boerum\n',
 'Abbott Daniel, 25 Front\t[Court\n',
 'Abbott Francis H, merchant 141 Front NYh Joralemon n\n',
 'Abbott H B, milkman Kent av north\n',
 'Abbott James, laborer Bond c State\t[Hoyt\n',
 'Abbott John D, merchant 295.5 Pearl NYh 362 Atlantic n\n',
 'Abbott Moses, furrier 74 Concord\n',
 'Abbott Samuel D, butcher Myrtle c Ja

## Divide records into different sections based on first letter

In [28]:
data[0] = 'A.'
import re

d = {}
value = []
i = 0
for line in data:
  line = line.strip()
  if line:
    if len(line) < 3: 
      # 1850 data is not split by '***' instead it is split by 'B., C., D.'
      if bool(re.match("([A-Z]+\.)", line)) == True:
        # save previous value
        if value:
          if value[0] == "'":
            value = value[1:]
          d[key] = value
        
        key = line
        value = []
    else:
      value.append(line)

In [30]:
d.items

<function dict.items>

## Combline lines that belong to the same record

Criteria:

1) First letter of next line not equal to the initial letter in this section

2) Special cases in the start of next line: B’klyn, B'way, E., W.

In [31]:
for key,value in d.items():
  initial = key[0]
  new_value = []
  n = len(value)
  i = 0
  
  while i < n:
    if i < n-1:
      if (value[i+1][0] != initial and value[i+1][0] != "*") or (value[i+1][0:2] in ["B’", "E.", "W."]) or (value[i+1][0:2] == ". "):
        if value[i][-1] == "-":
          s = value[i][:-1] + value[i+1]
        else:
          s = value[i] + ' ' + value[i+1]
        new_value.append(s)
        i += 2
      else:
        new_value.append(value[i])
        i += 1 
    else:
      if value[i][0] == initial:
        new_value.append(value[i])
      i += 1
  
  d[key] = new_value      

In [32]:
# remove "(see adv...."
for key,value in d.items():
  n = len(value)
  for i in range(n):
    ls = value[i].split()
    n2 = len(ls)
    for j in range(n2):
      if "see" in ls[j]:
        value[i] = ' '.join(ls[:j])
        break

## Identify business vs person records

Made an update on 27/03/20 to change the index notation.

* The first letter of Company names were being removed by the function
    * ZUCCA BROTHERS
    * ARNAULT PETER
    * APPELL JACOB
    * APPLETON’S AMERICAN CYCLOPG5DIA
    * BURTON A. B 

In [33]:
# add an index to each record
# in another dictionary, for each index specify whether it's business or person record
index = 1
d_bp = []
for key,value in d.items():
  n = len(value)
  for i in range(n):
    ls = value[i].split()
    if ls:
      if ls[0][0] == "*" or (len(ls[0]) > 1 and ls[0].isupper()): # Upper case as well
        value[i] = value[i][0:] # Changed from [0:] to [1:]
        d_bp.append([str(index), "b"])
      else:
        d_bp.append([str(index), "p"])
    
      value[i] = str(index) + "\t" + value[i]
      index += 1

## Save files

In [34]:
file = open("input/1850/nypl_1850_51_starred_clean.txt", "w")
for key,value in d.items():
  file.write(key + '\n')
  for record in value:
    file.write(record + '\n')
  file.write('\n')

file.close()

file2 = open("input/1850/index_business_person.txt", "w")
for item in d_bp:
  file2.write(item[0] + ' ' + item[1] + '\n')
  
file2.close()